termwiz/src/error.rs

```diff
@@ -0,0 +1,37 @@
+//! Error types.
+use thiserror::Error;
+
+/// Convenient return type for functions.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Main error type.
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Generic I/O error.
+    #[error("i/o: {0}")]
```

Comment:
I think it would save some bikeshedding if these all (except for `ImpossibleState`) just passed through to the underlying `Display` impl;

```
#[error("{0}")]
```

for each variant; that way we don't need to worry about the additional phrasing for the different variants.

Reply:
Sort of on the same line of thinking: which of these errors are you thinking of handling in code in consumers of `termwiz`?  There's kind of a lot of effort to produce all these different error variants, and in turn, that makes it more difficult to iterate on termwiz in the future.

I can understand not wanting to bleed `anyhow` into consuming code, but I'm also concerned about the fan-out of enum variants and the "fan-in" impact within the crate.

What do you think about reducing the scope of the error variants here so that there are fewer to manage both inside and outside of the `termwiz` crate?

The most basic way of looking at that is that this `Error` type here is largely opaque outside of the crate (just implements the error related traits).

Most of the error conditions raised during escape sequence parsing aren't really discrete conditions that the calling code can do anything about, so I don't think that having so many error variants is especially helpful.  Also: I do rather enjoy the convenience of `bail!`, `ensure!` and `anyhow!` macros for constructing errors.  It is certainly lazy, but it seems like "good lazy" given that distinguishing most (or all?) of these cases generally doesn't matter.

What do you think?

Reply:
Didn't really have a reason, I only mechanically converted the existing `bail!` etc messages with the goal to not discard any information, but more from a perspective of _developer_ debuggability (seemed annoying to delete all the bail! etc stringed info) rather than consumers necessarily caring about, say, the particular ioctl that was called. Generally, though:

 - getting io::Error information that I don't have to parse out of a string is useful, because often it can be "no permissions" which can be handled differently from "file doesn't exist" which is also different from "no space left on device" etc.
 - Fmt, Regex, Utf8, these all can collapse into a generic opaque type. Same for where they appear in sub error types.
 - The wrapped sub error types are mostly to distinguish these two broad categories to make it easier for cases that don't care about the more precise errors to glob those out without having a single "every error is a string" type.

Going at an extreme (not serious), we could well have a top-level

```rust
#[derive(Debug, Error)]
pub enum Error {
	#[error("{0}")]
	Wrapped(anyhow::Error),
}

impl From<anyhow::Error> for Error {
	fn from(err: anyhow::Error) -> Self {
		Self::Wrapped(err)
	}
}
```

and be mostly done, keeping everything with anyhow internally.

More seriously though, and without redoing everything, I could add a `Generic(anyhow::Error)` with a `From` impl as above to the existing enums, and you'd be able to use anyhow macros internally with abandon, then specialise to a variant if needed/requested. (That's how I did the conversion piece by piece in the first place.) Then I'd undo all of the specialised error conversions that were a bit overzealous (like all the Tty variants).

Reply:
Rather than `#[error("{0}")]` you should instead use `#[error(transparent)]` for errors that you want to pass through directly.

Similarly, errors fields marked as `#[source]` (or `#[from]` which implies source) shouldn't be included in the formatting.  The user of the error can walk the error chain if they want to know more, and if you include the source error in the formatting then users that *do* walk the chain will get the error printed twice, which looks ugly.

I made simliar changes in streampager [here](https://github.com/markbt/streampager/commit/44cae6d89b983903186f4c602d9c35fe97c503e5).

Reply:
Ah, good point. I'll change things around (tomorrow, it's late here).